Like block: Fix layout shift when loading

ntsekouras opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes https://github.com/Automattic/jetpack/issues/36056


Like block is causing a layout shift when loading. This PR addresses that by setting an initial zero height to the iframe and update it on load. At the same time (on iframe load) we hide the placeholder.

I've also update the `comments like` display to happen on load, but while in theory seems right to me, I'm not 100% sure  because I couldn't figure out where this renders. Can someone help me out with this about where should I be testing?

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?


## Does this pull request change what data or activity we track or use?
no

## Testing instructions:
1. Insert a Like block and observe that in the front-end there is no layout shift.







#### Before:
https://github.com/user-attachments/assets/676a41ee-937b-4cc6-8a51-adc84b57a41f

#### After:
https://github.com/user-attachments/assets/da758005-2ab1-4ae2-a6c7-9f2e050a93dd

ntsekouras committed:
Like block: Fix layout shift when loading

ntsekouras commented on projects/plugins/jetpack/extensions/blocks/like/like.php:
```diff
@@ -106,7 +106,7 @@ function render_block( $attr, $content, $block ) {
 	$wrapper = sprintf( 'like-post-wrapper-%1$d-%2$d-%3$s', $blog_id, $post_id, $uniqid );
 
 	$html = "<div class='sharedaddy sd-block sd-like jetpack-likes-widget-wrapper jetpack-likes-widget-unloaded' id='" . esc_attr( $wrapper ) . "' data-src='" . esc_attr( $src ) . "' data-name='" . esc_attr( $name ) . "' data-title='" . esc_attr( $title ) . "'>"
-		. "<div class='likes-widget-placeholder post-likes-widget-placeholder' style='height: 55px;'><span class='button'><span>" . esc_html__( 'Like', 'jetpack' ) . "</span></span> <span class='loading'>" . esc_html__( 'Loading...', 'jetpack' ) . '</span></div>'
```
I removed this button as it seemed like a leftover or a result of quick copy/paste code..

ivan-ottinger replied:
Yes, this looks like unused code indeed. 👍🏼

jsnajdr replied:
The removed button was probably part of the "old layout". There's been a recent "new layout" implementation sometimes 2 years ago, turned on by feature flags and CSS classes. In the new layout the button is apparently always hidden so it can be removed from the markup. We'll see more of these little cleanup opportunities for a long time 🙂 

It would be also nice to update the ellipsis character (...) in the "Loading..." label. Instead of three dots use the special unicode char. I see in the Jetpack codebase that the new string is already used at many places, so should be already translated.

ntsekouras commented on projects/plugins/jetpack/modules/likes/queuehandler.js:
```diff
@@ -170,14 +165,6 @@ function JetpackLikesMessageListener( event ) {
 			break;
 		}
 
-		case 'showCommentLikeWidget': {
```
I'd appreciate some guidance where/how I can test this. In theory it seems there wouldn't be a problem due to [this code](https://github.com/Automattic/jetpack/pull/42361/files#diff-39b3a028d7cc7e1ecf123770f1cc2c6bd3ab1739c3906c06f47145b4eb5fe96bR434), but we need to be sure.

ivan-ottinger replied:
This is not clear to me either. When we worked on the Like button back in Apex, we focused on the post likes only (as opposed to comment likes).

Maybe @jsnajdr will have some idea (?) 🤔

jsnajdr replied:
You can test this on an Atomic or Jetpack site (e.g., Jurrassic Ninja) where you have the Jetpack Beta plugin and are able to install Jetpack build from this PR branch.

Posts with comments will display both Post Likes and also Comment Likes on each comment. I have some screenshots in a wpcom PR (wpcom-gh176570), you can see there how it looks like.

Post and Comment Likes can be enabled/disabled separately, maybe that's the reason why you don't see the comment likes. There are two checkboxes in two sections of Jetpack Settings:

<img width="936" alt="Screenshot 2025-03-12 at 10 18 50" src="https://github.com/user-attachments/assets/cfb0df8e-14d9-4537-b5bc-b335ccad0371" />

I plan to test this PR today, I'll have more feedback then.

ivan-ottinger replied:
Thank you for chiming in, Jarda! I appreciate your input.

ntsekouras replied:
Thanks for the context @jsnajdr! I have the setting for comment likes enabled and it's probably this one:
<img width="357" alt="Screenshot 2025-03-12 at 4 02 21 PM" src="https://github.com/user-attachments/assets/d73fe481-2250-4138-b912-a2cf3cf2af4e" />

I had added breakpoints there though and it wouldn't stop in my testing though..

jsnajdr replied:
Where are you setting the breakpoint? The workflow while loading is:
1. `queuehandler.js` in the top frame observes which like widgets are in view or have just scrolled in. In such a case it creates (invisible) iframe with the widget, and when it loads (there is only an empty placeholder inside) it sends a `loadLikeWidget` message to the `likes-master` frame.
2. The `likes-master` frame populates the `iframe` with content, after fetching the likes async from REST. Then it sends a `showLikeWidget` message back to the top frame. The top frame will then remove the "Loading" placeholder (which is owned by the top frame).

Someone (probably the top frame) will also make the loaded iframe visible at the right moment by removing `display:none`.

The comment likes should work exactly the same way, only the messages are slightly different (either name is different or one of the params is different).

I'm also testing this right now, hopefully will have some feedback soon.

ntsekouras replied:
My breakpoint was inside [showCommentLikeWidget case](https://github.com/Automattic/jetpack/pull/42361/files#diff-39b3a028d7cc7e1ecf123770f1cc2c6bd3ab1739c3906c06f47145b4eb5fe96bL173) for comments.

jsnajdr replied:
When I set a breakpoint on `showCommentLikeWidget`, it's triggered:

<img width="1038" alt="Screenshot 2025-03-12 at 15 54 47" src="https://github.com/user-attachments/assets/3c302d83-447a-4471-b4e2-f368dfb066fe" />

Maybe you're in the wrong `queuehandler.js` file? 🙂 There are two, one from Jetpack, one from `mu-plugins/likes`.

Also, the Comment Like widget might not need this patch -- does it really have the same layout shift? The star icon is part of the "wrapper", and the "Loading" placeholder is neatly aligned with the text-only content of the iframe widget.

ntsekouras replied:
I will look into the breakpoint thing..

For now I removed the patch for comment likes

ivan-ottinger reviewed (dismissed):
Thank you for the PR review request, Nik.

I can confirm the proposed changes fix the issue. There's no layout shift anymore. 👌🏼

I did not observe any regressions either.

We may need to apply similar fix to the `queuehandler.js` on the WPCOM side as well: fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Szh%2Qcyhtvaf%2Syvxrf%2Sdhrhrunaqyre.wf%3Se%3Q2os784s6-og - as far I can see the same issue is present on Simple sites too.

---

Generally, it might be worth checking with @jsnajdr as well, since he worked on some Like button changes recently.

ntsekouras commented:
>We may need to apply similar fix to the queuehandler.js on the WPCOM side as well: uggcf%3N%2S%2Suers.yv%2S%3Suggcf%3N%2S%2Sbcratebx.n8p.pbz%2Sfbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Szh%2Qcyhtvaf%2Syvxrf%2Sdhrhrunaqyre.wf%3Se%3Q2os784s6-og - as far I can see the same issue is present on Simple sites too.

That's interesting and explains why I couldn't test it in simple sites properly. So in simple sites we override this file I guess, so I opened another 176735-ghe-Automattic/wpcom. What kind of syncing between the two files is needed though? 🤔 

For example in the same code path in Jetpack as here in wpcom 176735-ghe-Automattic/wpcom/files#diff-cfaa555eedf0a4c4fa5e18e05fee801993c49d447ae1a6516b3104f5502dcd51L159, there is less code - `if ( data.total > 0 ) {...` is missing.

ivan-ottinger commented:
> > We may need to apply similar fix to the queuehandler.js on the WPCOM side as well: uggcf%3N%2S%2Suers.yv%2S%3Suggcf%3N%2S%2Sbcratebx.n8p.pbz%2Suggcf%3N%2S%2Suers.yv%2S%3Suggcf%3N%2S%2Sbcratebx.n8p.pbz%2Sfbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Szh%2Qcyhtvaf%2Syvxrf%2Sdhrhrunaqyre.wf%3Se%3Q2os784s6-og - as far I can see the same issue is present on Simple sites too.
> 
> That's interesting and explains why I couldn't test it in simple sites properly. So in simple sites we override this file I guess, so I opened another 176735-ghe-Automattic/wpcom. What kind of syncing between the two files is needed though? 🤔
> 
> For example in the same code path in Jetpack as here in wpcom 176735-ghe-Automattic/wpcom/files#diff-cfaa555eedf0a4c4fa5e18e05fee801993c49d447ae1a6516b3104f5502dcd51L159, there is less code - `if ( data.total > 0 ) {...` is missing.

If nothing has changed in the meantime, those two files are not synced. When we worked on the Like button redesign project, it was already set up that way and there have been some differences between the files.

There were discussions on unifying the logic, but due to time constrains it hasn't progressed further.

jsnajdr commented:
> That's interesting and explains why I couldn't test it in simple sites properly.

Simple sites use a legacy implementation in `wp-content/mu-plugins/comment-likes`. Use this patch of mine to disable them and enable the "Jetpack" ones: #172682

> What kind of syncing between the two files is needed though?

Instead of syncing, we should remove `wp-content/mu-plugins/likes` and enable the corresponding Jetpack module instead. Any code differences are a consequence of neglected syncing, it's not an intent.

That's a topic for a separate PR, and I think now we have gathered enough motivation to really do it 🙂

ntsekouras committed:
bring back `showCommentLikeWidget` case

jsnajdr commented:
I figured out why the layout shift is happening and I think the fix should be different.

Both the placeholder and the iframe are boxes with `height: 55px` inside a wrapper element. When the widget is loading, they are both displayed at the same time. And when the load is finished, the placeholder is hidden with `display: none`. That causes the layout shifts.

There are three ways how the widget can be inserted:
1. Simple site will put a "post flair" element after the post content: `<div id="jp-post-flair" class="sharedaddy sd-like-enabled" />`. There's a `mu-plugin` for that that runs only on Simple sites. The Like widget is inside this flair `div`. Note the flair's CSS classes, they are important.
2. Atomic/Jetpack site will put the Like widget after the post content, but there's no flair element.
3. The Like block is rendered wherever you put it, in a theme template or even inside post body, without any limitations.

Now, there is a CSS style for the Like widget that looks like this:
```css
div.sharedaddy.sd-like-enabled .sd-like .post-like-widget {
  position: absolute;
  top: 0;
}
```
This style makes the iframe (`.post-like-widget`) inside the wrapper (`.sd-like`) overlap the placeholder and prevents the layout shift, because the absolutely-positioned element is outside the layout flow. And because the wrapper has fixed height, it doesn't mind that its content has effectively zero height, layout-wise.

However, this style is active only inside the post flair, i.e., only on Simple sites! If the Likes widget is rendered by a block, or on Atomic site, there is a layout shift.

We need to remove the `div.sharedaddy.sd-like-enabled` prefix from the CSS rule. Then the styles that prevent the layout shift will be active in all circumstances.

Or we can completely rethink how the placeholder and the iframe are positioned and styled. We can set `display:none` to the iframe initially, and show it only after it's loaded. We can use the `jetpack-like-widget-*` CSS classes, they are currently nicely set and unset, but there is no styling associated with them. Example:
```css
.jetpack-like-widget-unloaded, .jetpack-like-widget-loading {
  .placeholder { display: block; }
  .frame { display: none; }
}
.jetpack-like-widget-loaded {
  .placeholder { display: none; }
  .frame { display: block; }
}
```
This always displays only one of the elements, without any race conditions (caused by cross-window async messages) where both are displayed.

ntsekouras committed:
css approach

ntsekouras commented:
Thank you @jsnajdr for your thorough investigation. A css approach as you suggested makes more sense and I wasn't sure why it wasn't preferred in the first place.

I updated the code and seems to work well. By having the `iframe` selector I also removed the `showCommentLikeWidget` again.

ntsekouras committed:
use ellipsis char

jeherve reviewed (dismissed):
This tests well for me. 👍

jsnajdr commented on projects/plugins/jetpack/modules/likes/style.css:
```diff
@@ -136,11 +136,23 @@ div.sd-box {
 .comment-likes-widget {
 	margin: 0;
 	border-width: 0;
-	display: block;
```
This `display: block` is IMO quite important, although removing it maybe fixes part of the layout shift. Without this, the `iframe` will have `display: inline` by default. But visually it's a block. Switching between `inline` and `block` changes layout (vertical spacing) because of how `line-height` works and how inline content is aligned to the baseline.

ntsekouras replied:
This is removed from here because we add the same rule [above](https://github.com/Automattic/jetpack/pull/42361/files#diff-fda7012c269c04c6e9c61cda097274dedf76a96b5fc1861bfc97fa48d95e9738R142).

And with this [one](https://github.com/Automattic/jetpack/pull/42361/files#diff-fda7012c269c04c6e9c61cda097274dedf76a96b5fc1861bfc97fa48d95e9738R148), the change is between `display:none` to `display:block`.

I removed it from here because:
1. I added above as I mentioned
2. I'd have to add more css specificity to override this rule, which I think it's not needed.

Am I missing something? 🤔

jsnajdr replied:
Thanks for the clarification, my concern was that `display:block` gets completely removed, but that turns out not to be true. :+1:

jsnajdr commented on projects/plugins/jetpack/modules/likes/queuehandler.js:
```diff
@@ -170,14 +165,6 @@ function JetpackLikesMessageListener( event ) {
-				placeholder.style.display = 'none';
-			}
-			break;
-		}
```
I'd recomment to keep the `showCommentLikeWidget` case although it will be empty. I believe it will be soon populated again.

Here's why: currently the `jetpack-likes-widget-loaded` class is set a bit too soon, after the like iframe loads (`load` event listener). But at this moment the iframe doesn't have any content yet. The `likes-iframe` and its `likes-rest-nojquery` script will add DOM markup there, after loading the likes from REST, and only then it will send the `show(Comment)LikeWidget` message back to `queuehandler`. And it's the handler for this message that should set the `jetpack-likes-widget-loaded`.

We should do this in one of followup PRs and then `showCommentLikeWidget` will be handled again.

ntsekouras replied:
I'm not very fond of keeping the empty block, but I get your point and added a comment why it's there, linking to this thread.

jsnajdr reviewed (dismissed):
This is a good step forward, although I have ideas for further improvements. But this feature is very entangled, so let's untangle it one step at a time.

ntsekouras committed:
keep empty `showCommentLikeWidget`
